Sprint 17

brozozowska opened the pull request:
Задания спринта 17 выполнены

brozozowska committed:
feat(localization): перенёс склонение «день» в .stringsdict и удалил метод dayWord(for:)

brozozowska committed:
feat(localization): приложение полностью локализовано на en и ru
дополнительно убрал из кода использование rawValue перечисления WeekDay, для корректной работы локализации

brozozowska committed:
feat: реализован поиск по созданным трекерам

brozozowska committed:
fix: изменено отображение и анимации строки поиска в соответствии с макетом

brozozowska committed:
feat: реализована тёмная тема на всех экранах

brozozowska committed:
feat: реализовано редактирование созданного трекера

brozozowska committed:
feat: реализован экран статистики
обновлены настройки НавБара по умолчанию для больших и маленьких тайтлов
добавлен новый инициализатор WeekDay(from:, calendar:) для переиспользования в StatisticsViewController и TrackersViewController
заменена картинка-заглушка при пустой статистике

brozozowska committed:
feat: добавлена новая картинка и лейбл для состояния, когда поиск по имени не находит ни одного трекера

brozozowska committed:
fix: исправлена ошибка удаления трекера
если трекер повторялся в другие дни недели, то после его удаления и перезапуска приложения трекер появлялся вновь

brozozowska committed:
fix: исправлена ошибка удаление категории
если удаленная категория была записана хоть в одном трекере, то после перезапуска приложения категория появлялась вновь

brozozowska committed:
feat: добавлена кнопка Фильтр на экран трекеров и добавлен экран со списком фильтров
переименована ячейка категории для переиспользования в фильтрах
добавлен отступ при прокрутке коллекции трекеров, чтобы новая кнопка не загораживала трекеры

brozozowska committed:
feat: добавлена логика отображения трекеров в зависимости от выбранного фильтра

brozozowska committed:
feat: при активном фильтре вокруг кнопки появляется градиентная граница
градиентная граница вынесена из StatisticsCell в отдельный файл

brozozowska committed:
test: добавлены скриншот-тесты экрана трекеров

brozozowska committed:
feat: добавил сбор аналитики на экран трекеров

NaR0RoW commented on Tracker/Screens/Statistics/StatisticsViewController.swift:
```diff
@@ -7,25 +7,77 @@
     }()
     
+    // MARK: - Stores & Service
+    private let trackerStore = TrackerStore()
+    private let recordStore = TrackerRecordStore()
+    private lazy var statisticsService = StatisticsService(
```
Было бы лучше если б контроллер инициализировался от сервиса или можно добавить специальный метод устанавливающий свойство. В будущем это облегчило бы тестирование, да и делает код более гибким и готовым к переиспользованию

brozozowska replied:
Обновил — вынес зависимости из наружу, добавил протокол для сервиса, добавил convenience init() и designated init() в контроллер

NaR0RoW commented on Tracker/Screens/Trackers/FiltersViewController.swift:
```diff
@@ -0,0 +1,132 @@
+    
+    // MARK: - UI Constants
+    private enum UIConstants {
+        static let TopOffset: CGFloat = 24
```
Название констант должно начинаться со строчной буквы

brozozowska replied:
Исправил

NaR0RoW commented on Tracker/Screens/Trackers/NewCategoryViewController.swift:
```diff
@@ -119,7 +121,14 @@ final class NewCategoryViewController: UIViewController, UITextFieldDelegate {
         doneButton.isEnabled = enabled
-        doneButton.backgroundColor = enabled ? .black : .lightGray
+        
+        if enabled {
+            doneButton.backgroundColor = .label
+            doneButton.setTitleColor(.systemBackground, for: .normal)
+        } else {
+            doneButton.backgroundColor = .lightGray
+            doneButton.setTitleColor(.white, for: .normal)
+        }
```
Лучше использовать тернарный оператор:
``` swift
doneButton.backgroundColor = enabled ? .label : .lightGray
doneButton.setTitleColor(enabled ? .systemBackground : .white, for: .normal)
```

NaR0RoW commented on Tracker/Screens/Trackers/NewTrackerViewController.swift:
```diff
@@ -479,7 +554,13 @@ final class NewTrackerViewController: UIViewController, NewScheduleViewControlle
         
         createButton.isEnabled = isValid
-        createButton.backgroundColor = isValid ? .black : .lightGray
+        if isValid {
+            createButton.backgroundColor = .label
+            createButton.setTitleColor(.systemBackground, for: .normal)
+        } else {
+            createButton.backgroundColor = .lightGray
+            createButton.setTitleColor(.white, for: .normal)
+        }
```
Лучше использовать тернарный оператор

brozozowska replied:
Исправлено

NaR0RoW commented on Tracker/Screens/Trackers/TrackersViewController.swift:
```diff
@@ -16,6 +16,28 @@ final class TrackersViewController: UIViewController, NewTrackerViewControllerDe
+    }
+    
+    // MARK: - Analytics
+    private let analytics = AnalyticsService()
```
`AnalyticsService` можно сделать синглтоном

NaR0RoW commented on TrackerTests/__Snapshots__/TrackerTests/testTrackersScreenDark.1.png:
Генерируемые сторонними библиотеками файлы и папки лучше выносить в [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files)

brozozowska replied:
Согласен, но не для скриншот-тестов. Мне кажется эталонные скрины для тестов имеет смысл передавать с коммитом, так как это часть тестов

brozozowska commented on Tracker.xcodeproj/xcuserdata/srozov.xcuserdatad/xcschemes/xcschememanagement.plist:
Да, проверил самый первый коммит, xcuserdata в него попала. А .gitignore появился только во втором коммите. Удалю в следующем коммите, спасибо.

brozozowska committed:
chore: удалил xcuserdata из репозитория

brozozowska committed:
refactor: исправлены мелкие замечания код-ревью
удалены лишние ассеты
убраны лишние return из однострочных выражений
название константы теперь написано со строчной буквы

brozozowska committed:
refactor: конструкции if/else заменены тернарными операторами для улучшения читаемости

brozozowska committed:
refactor: вынесены зависимости из StatisticsViewController наружу
добавлен StatisticsServiceProtocol для сервиса
добавлен convenience init() с дефолтными зависимостями
добавлен designated init() для задания новых зависимостей

brozozowska committed:
refactor: AnalyticsService сделан синглтоном

brozozowska commented on Tracker/Models/TrackerFilter.swift:
```diff
@@ -0,0 +1,28 @@
+        case .all:
+            return NSLocalizedString("filters.all", comment: "All trackers")
+        case .today:
+            return NSLocalizedString("filters.today", comment: "Trackers for today")
```
Да, мигрировал локализацию в проекте на string catalog

brozozowska committed:
refactor(localization): локализация проекта смигрирована на string catalogs

brozozowska committed:
refactor(localization): прямые вызовы NSLocalizedString заменены на обёртку Localizable

brozozowska commented on Tracker/Models/TrackerFilter.swift:
```diff
@@ -0,0 +1,28 @@
+    var title: String {
+        switch self {
+        case .all:
+            return NSLocalizedString("filters.all", comment: "All trackers")
```
Сделано. Заменил прямые вызовы NSLocalizedString на обёртку Localizable
